governance/proposals/up/009-add-uniswap-liquidity.js

```diff
@@ -0,0 +1,217 @@
+  const amount1Min = calculateMinimumAmount(amount1DesiredRaw)
+
+  return {
+    params: {
+      token0: pool.token0.address,
+      token1: pool.token1.address,
+      fee: pool.fee,
+      tickLower: FULL_RANGE_LOWER_TICK,
+      tickUpper: FULL_RANGE_UPPER_TICK,
+      amount0Desired: amount0DesiredRaw,
+      amount1Desired: amount1DesiredRaw,
+      amount0Min,
+      amount1Min,
+      recipient: TIMELOCK_ADDRESS,
+      deadline: DEADLINE,
+    },
+    requiredAmounts: {
+      token0: amount0DesiredRaw,
+      token1: amount1DesiredRaw,
+    },
+  }
+}
+
+const buildProposalDescription = ({
+  ethAmount,
+  needsWrapping,
+  needsWethApproval,
+  needsUpApproval,
+}) => `Add Liquidity to UP/WETH Uniswap Pool
+
+This proposal deposits ${ethAmount} ETH and the corresponding amount of UP
+tokens into a full-range position in the UP/WETH Uniswap V3 pool on Base. It will:
+
+1. Improve liquidity and price stability for UP  
+2. Generate fee income for the DAO  
+3. Make it easier for users to trade UP  
+
+The proposal executes these steps:
+${[
+  needsWrapping
+    ? 'Wrap ETH to WETH (treasury WETH balance is insufficient)'
+    : null,
+  needsWethApproval ? 'Approve WETH for the position manager' : null,
+  needsUpApproval ? 'Approve UP tokens for the position manager' : null,
+  'Create the full‑range liquidity position',
+]
+  .filter(Boolean)
+  .map((step, i) => `${i + 1}. ${step}`)
+  .join('\n')}
+
+The position will be owned by the DAO treasury (Timelock contract) and
+can be managed through future governance proposals.`
+
+const WETH_ABI = ['function deposit() payable']
+const POSITION_MANAGER_ABI = [
+  'function mint((address token0, address token1, uint24 fee, int24 tickLower, int24 tickUpper, uint256 amount0Desired, uint256 amount1Desired, uint256 amount0Min, uint256 amount1Min, address recipient, uint256 deadline)) returns (uint256 tokenId, uint128 liquidity, uint256 amount0, uint256 amount1)',
+]
+
+module.exports = async () => {
+  const { tokens } = await getNetwork()
+  const { chainId } = await ethers.provider.getNetwork()
+  const {
+    uniswapV3: { positionManager: positionManagerAddress },
+  } = networks[chainId]
+
+  const weth = tokens.find((token) => token.symbol === 'WETH')
+  const up = tokens.find((token) => token.symbol === 'UP')
+  const wethContract = await ethers.getContractAt(ERC20_ABI, weth.address)
+  const upContract = await ethers.getContractAt(ERC20_ABI, up.address)
+  const { params, requiredAmounts } = await calculatePosition(ETH_AMOUNT)
+  const wethBalance = await wethContract.balanceOf(TIMELOCK_ADDRESS)
+  const ethAmount = ethers.parseEther(ETH_AMOUNT)
+  const needsWrapping = wethBalance < ethAmount
+
+  const [wethAllowance, upAllowance] = await Promise.all([
+    wethContract.allowance(TIMELOCK_ADDRESS, positionManagerAddress),
+    upContract.allowance(TIMELOCK_ADDRESS, positionManagerAddress),
+  ])
+
+  const needsWethApproval = wethAllowance < ethAmount
```

Comment:
same here, calculate directly the amount of tokens to approve

Reply:
@blahkheart ^^